src/main/java/org/junit/internal/StackTraces.java

```diff
@@ -0,0 +1,203 @@
+            List<String> stackTraceLines, boolean hasCause) {
+        State state = State.PROCESSING_OTHER_CODE;
+        int linesToInclude = stackTraceLines.size();
+        for (String stackTraceLine : asReversedList(stackTraceLines)) {
+            state = state.processLine(stackTraceLine);
+            if (state == State.DONE) {
+                List<String> trimmedLines = stackTraceLines.subList(0, linesToInclude);
+                if (!hasCause) {
+                    return trimmedLines;
+                }
+                List<String> copy = new ArrayList<String>(trimmedLines);
+                copy.add("\t..." + (stackTraceLines.size() - copy.size()) + "more");
+                return copy;
+            }
+            linesToInclude--;
+        }
+        return Collections.emptyList();
+    }
+
+    private static <T> List<T> asReversedList(final List<T> list) {
+        return new AbstractList<T>() {
+
+            @Override
+            public T get(int index) {
+                return list.get(list.size() - index - 1);
+            }
+
+            @Override
+            public int size() {
+                return list.size();
+            }
+        };
+    }
+
+    private enum State {
+        PROCESSING_OTHER_CODE {
+            @Override public State processLine(String line) {
+                if (isTestFrameworkStackTraceLine(line)) {
+                    return PROCESSING_TEST_FRAMEWORK_CODE;
+                }
+                return this;
+            }
+        },
+        PROCESSING_TEST_FRAMEWORK_CODE {
+            @Override public State processLine(String line) {
+                if (isReflectionStackTraceLine(line)) {
+                    return PROCESSING_REFLECTION_CODE;
+                } else if (isTestFrameworkStackTraceLine(line)) {
+                    return this;
+                }
+                return PROCESSING_OTHER_CODE;
+            } 
+        },
+        PROCESSING_REFLECTION_CODE {
+            @Override public State processLine(String line) {
+                if (isReflectionStackTraceLine(line)) {
+                    return this;
+                } else if (isTestFrameworkStackTraceLine(line)) {
+                    // This is here to handle TestCase.runBare() calling TestCase.runTest().
+                    return PROCESSING_TEST_FRAMEWORK_CODE;
+                }
+                return DONE;
+            } 
+        },
+        DONE {
+            @Override public State processLine(String line) {
+                return this;
+            } 
+        };
+
+        /** Processes a stack trace line, possibly moving to a new state. */
+        public abstract State processLine(String line);
+    }
+
+    private static final String[] TEST_FRAMEWORK_METHOD_NAME_PREFIXES = {
+        "org.junit.runner.",
+        "org.junit.runners.",
+        "org.junit.experimental.runners.",
+        "org.junit.internal.",
+        "junit.",
+        "org.apache.maven.surefire."
```

Comment:
Thanks; I'll try to find time to add this to the test cases this weekend. It's possible that Gradle doesn't use the code I'm changing when it prints out test failures or produces test XML.